engine/buff/buff.cpp

```diff
@@ -829,6 +831,76 @@ buff_t* buff_t::set_duration_multiplier( double multiplier )
   return this;
 }
 
+buff_t* buff_t::set_time_modifier( double modifier )
+{
+  if ( s_data->flags( spell_attribute::SX_IGNORE_FOR_MOD_TIME_RATE ) )
+    return this;
+
+  modifier = std::max( modifier, -99.0 ); // Limit slow down to 100x slower
+
+  if ( modifier == time_modifier )
+    return this;
+
+  auto old_modifier = time_modifier;
+  time_modifier = modifier;
+
+  if ( current_stack <= 0 || expiration.empty() )
+  {
+    return this;
+  }
+
+  auto ratio = ( old_modifier + 100.0 ) / ( time_modifier + 100.0 );
+
+  // Slowing down the clock, expiry will be moved into the future
+  if ( time_modifier < old_modifier )
+  {
+    for ( auto exp : expiration )
+    {
+      auto new_remains = exp->remains() * ratio;
+      exp->reschedule( new_remains );
+
+      if ( player )
+        sim->print_log( "{} increases {} expiry by a factor of {}. New expiration: {}", *player, *this, ratio,
+                      exp->occurs() );
+    }
+  }
+  // Speeding up the clock, expiry moved closer
+  else
+  {
+    auto old_expirations = new std::vector<event_t*>();
+    old_expirations->swap( expiration );
+
+    for ( auto old_expiration : *old_expirations )
+    {
+      auto new_remains = old_expiration->remains() * ratio;
+      event_t::cancel( old_expiration );
+
+      auto exp = make_event<expiration_t>( *sim, this, new_remains );
+      expiration.push_back( exp );
+
+      if ( player )
+        sim->print_log( "{} decreases {} expiry by a factor of {}. New expiration: {}", *player, *this, ratio,
+                      exp->occurs() );
+    }
+    old_expirations->erase( old_expirations->begin(), old_expirations->end() );
+
+  }
+
+  return this;
+}
+
+buff_t* buff_t::pause()
```

Comment:
I'm not convinced these functions should be part of the public interface of buff_t. Especially when pause() doesn't really pause anything, just slows it down by A TON.

Reply:
Correct! But in the verbage of talents like underhanded upper hand, the text says the buff duration "does not decrease", when in reality what's happening is that this style of time modifier is being applied so long as the other buff is active. It seems likely this is how buff pausing effects will be implemented in the future (via this fake pause via 100x slowdown), since UI addons and the like handle it better than a true pause, thanks to already using the time modifier for _accelerated_ cooldown scenarios presented by some bosses since legion, so a function for that behavior seems useful. At least it was useful for me, since I had to call it a fair bit because the aura-applying infrastructure for buffs doesn't seem to handle _un_-applying auras too well, so seems unsuitable for auras applying temporary changes to other auras (like what the serverside scripts for underhanded upper hand seem to do with its' background spells).

Basically I just found it useful and think it'll probably be useful in the future, too. If not in the base buff type, I'd probably add it into the rogue buff base, because it seems cleaner, intent-wise, than calling `set_time_modifier` directly a bunch, and rogue needs it on at least 3 buffs.

Reply:
I don't mind the idea in principle, it's just that right now, the only example we have is the new rogue stuff. If you make a shortcut in the rogue module that's perfectly fine. If they reuse it somewhere else, that's when I think we should be considering making it part of the buff_t interface.

I guess the bigger issue for me is that this function conveys the wrong idea to devs not familiar with the rogue rework, especially since we sometimes use buff_t for stuff that doesn't have a corresponding buff in game, where you might want to actually pause a buff rather than slow it down by a lot. If this way of "pausing" buffs becomes established as a normal thing they use, then I don't think there would be any risk of confusion.